zerver/tests/test_auth_backends.py

```diff
@@ -2820,6 +2857,30 @@ def test_github_unverified_email_with_existing_account(self) -> None:
             "warning",
         )])
 
+    def test_social_auth_email_change_unregistered_user_login(self) -> None:
+        # Overriden as user is taken to choose email screen and
+        # it needs to be handled by send `expect_choose_email_screen`
+        # to `social_auth_test`.
```

Comment:
is it not possible to write the `do_email_change` function in such a way that this override won't be needed?

Reply:
We'd need to have conditionals to check if is github auth and add some code to check that the user doesn't exist which means that it most probably is this particular test for unregistered user, then we need to set `expect_choose_email_screen`. I think this would clutter up `do_email_change`. One other thing we could do is send some param `unregistered_user=True` to avoid some code and add an `if backend == GitHubAuthBackend and unregistered` in the `do_email_change` and then set `expect_choose_email_screen` to True. But this param would remain useless for other auth flows and become confusing to someone reading the code if we introduce an argument for `do_change_email` for the purpose of a single test.

Both of those options seemed not optimal to me and had to write this overridden function.

Reply:
> then we need to set expect_choose_email_screen. I think this would clutter up do_email_change

That sounds like simply an extra argument and a 2-liner `if` case in `do_change` though I think? Or would it be more clutter than that

Reply:
> That sounds like simply an extra argument and a 2-liner `if` case in `do_change` though I think?

Yes, if we give an extra argument it's just a simple if in `do_change`. But I'm not sure if it's fine to send an arg just for Github where the test just doesn't run in social auth flow. Is it fine?

Reply:
> That sounds like simply an extra argument and a 2-liner `if` case in `do_change` though I think?

Yes, if we send an argument it's just a simple `if`, but not sure if it's fine to send it as an arg from the test only for GIthub auth where it is being run in 2 other test classes too. Is it fine?

Reply:
we could do this in `do_email_change` to avoid sending an arg, what is your opinion on this. (This seems like clutter to me but maybe not)
```diff
    def do_email_change(self, user_profile: UserProfile, action_key: str) -> HttpResponse:
         email = user_profile.delivery_email
         name = user_profile.full_name
 
+        expect_choose_email_screen = False
+        if self.backend == GitHubAuthBackend:
+            try:
+                get_user_by_delivery_email(email, user_profile.realm)
+            except UserProfile.DoesNotExist:
+                # The test is testing unregistered user logging in
+                # to change email. In such case, github auth takes to
+                # choose email screen.
+                expect_choose_email_screen = True
+
         account_data_dict = self.get_account_data_dict(email=email,
                                                        name=name)
 
         result = self.social_auth_test(account_data_dict, subdomain='zulip',
-                                       action_key=action_key)
+                                       action_key=action_key,
+                                       expect_choose_email_screen=True)
```